### PR TITLE
Refactor GatewayService

### LINKED ISSUE #89310: Clean up GatewayService
`GatewayService` implements logic for delaying state recovery in the cluster until enough data nodes have joined and/or some timeout has occurred. The implementation is pretty old and doesn't make use of much modern infrastructure for doing this kind of thing, making it pretty hard to follow and to improve (see e.g. #86588). We should clean it up and simplify it.

### PR DESCRIPTION
**Pull request by ywangd**
This PR refactors GatewayService with the goal to make it easier to add new features.

Resolves: #89310

### CODE REVIEW

**Comment by ywangd** [comment]
@DaveCTurner This is my first attempt to "clean up" the GatewayService class. I tried to work with [your suggestion](https://github.com/elastic/elasticsearch/issues/89310#issuecomment-1737042380) to use `SubscribableListener` and remove the boolean flags. But I am not sure whether this matches what you had planned especially around

> one-shot semantics and make things sensitive to the master term 

I am open to any suggestions and comments. Once we are happy with the overall approach. I also plan to add a few more tests. Thanks!

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java [hunk 1]
```diff
@@ -141,77 +147,92 @@ public void clusterChanged(final ClusterChangedEvent event) {
...
-                        }
-                    }
-                }, recoverAfterTime, threadPool.generic());
+        if (thisRecoveryPlanned == null) {
+            logger.debug("state recovery is in progress");
+            return;
+        }
+
+        // This node is ready to schedule state recovery
+        thisRecoveryPlanned.addListener(new ActionListener<>() {
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/gateway/GatewayService.java:176]: I wanted to use `andThen` but it does not expose exception to the consumer while we need the exception here to handle timeout.

**Comment by DaveCTurner** [comment]
Certainly nicer already yes. Just playing around with ideas here, but I'd like to try having an object which represents roughly "a pending state recovery in master term _t_". The master term always increases, and if we need to retry then that'll always be in a later term so we'll get a new one of these pending-state-recovery objects. Does that make sense?

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java [hunk 2]
```diff
@@ -141,77 +145,112 @@ public void clusterChanged(final ClusterChangedEvent event) {
...
+        }
+
+        // At this point, we know the state is not recovered and this node is qualified for state recovery
+        // But we still need to check whether a previous one is running already
+        final long currentTerm = state.term();
+        if (pendingStateRecovery.term < currentTerm) {
+            // Always start a new state recovery if the master term changes
+            // If there is a previous one still waiting, both will run but at most one of them will
+            // actually make changes to cluster state
+            pendingStateRecovery = new PendingStateRecovery(currentTerm);
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/gateway/GatewayService.java:164]: Here the code always schedule a new state recovery if it sees a new term (and the cluster state does need recovery). Not sure if this is what you mean by

> I suggested better if we started a new timeout when a new master is elected

This code does not try to cancel the previous pending state recovery since it can only be a best effort. The task may have already been submitted when we try to cancel it. Please let me know if you disagree.

**Comment by ywangd** [comment]
> I'd like to try having an object which represents roughly "a pending state recovery in master term t". The master term always increases, and if we need to retry then that'll always be in a later term so we'll get a new one of these pending-state-recovery objects.

I pushed 571db2defae5539764952f32b63a6dcea6900158 as an attempt to implement this idea. Please let me know whether it makes sense. Thanks!

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java [hunk 3]
```diff
@@ -131,8 +129,14 @@ public void clusterChanged(final ClusterChangedEvent event) {
         }
 
         final ClusterState state = event.state();
+        final DiscoveryNodes nodes = state.nodes();
 
-        if (state.nodes().isLocalNodeElectedMaster() == false) {
+        if (nodes.getMasterNodeId() == null) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/GatewayService.java:134]: I think this is already covered by the check on `nodes.isLocalNodeElectedMaster()`.

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java [hunk 4]
```diff
@@ -141,77 +145,112 @@ public void clusterChanged(final ClusterChangedEvent event) {
             return;
         }
 
-        final DiscoveryNodes nodes = state.nodes();
-        if (state.nodes().getMasterNodeId() == null) {
-            logger.debug("not recovering from gateway, no master elected yet");
-        } else if (recoverAfterDataNodes != -1 && nodes.getDataNodes().size() < recoverAfterDataNodes) {
+        if (recoverAfterDataNodes != -1 && nodes.getDataNodes().size() < recoverAfterDataNodes) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/GatewayService.java:148]: Maybe move this condition into the per-term check too?

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/gateway/GatewayService.java:164]: Yep that's the direction of which I was thinking indeed. Looking better.

I think I'd like the `RecoverStateUpdateTask` to remember the corresponding term with which it was registered, and verify that the term hasn't changed before it does anything. That way we don't need to worry about spurious recoveries from older terms that are no longer correct.

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java [hunk 5]
```diff
@@ -141,77 +145,112 @@ public void clusterChanged(final ClusterChangedEvent event) {
...
-                            runRecovery();
-                        }
-                    }
-                }, recoverAfterTime, threadPool.generic());
+        PendingStateRecovery(long term) {
+            this.term = term;
+        }
+
+        void maybeStart(int dataNodeSize) {
+            final SubscribableListener<Void> thisRecoveryPlanned;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/GatewayService.java:180]: On reflection doing this with a `SubscribableListener` actually seems more awkward than just sticking with the original `threadPool.schedule`.

**Review by DaveCTurner — COMMENTED** [review]
I like the direction. Left a few more suggestions.

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java (hunk 3 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/gateway/GatewayService.java]: Yep I kept it because it has a separate logging message. I have now moved it inside `isLocalNodeElectedMaster` check to retain the logging message.

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java (hunk 4 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/gateway/GatewayService.java]: I moved the check inside the new `PendingStateRecovery` class before scheduling the recovery. I assume this is what you mean by "per-term", it is *not* about checking it inside `ClusterStateUpdateTask#execute`. The former allows us to keep the same semantics as of today, i.e. no action at all until the required number of data nodes is met.

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java (hunk 5 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/gateway/GatewayService.java]: Sure I changed it to use just `threadPool.schedule`. It might make  `SubscribableListener` more suitable for this use case if it does not have to use Exception to signal timeout.

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java (hunk 2 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/gateway/GatewayService.java]: That's a good idea. I added the term check.

**Comment by ywangd** [comment]
I made a rather large change based on last round's feedback. The code no longer needs synchronization. It now uses an AtomicReference to hold the current pending recovery. The task still cleans itself up when finishes either successfully and exceptionally. I personally think the code looks simpler with the latest change.

**Comment by DaveCTurner** [comment]
Right yes this is looking even better. But we are still doing some kinda complex state management within each term via `resetState()`. I think it'd be yet simpler to keep hold of the same `PendingStateRecovery` instance for the entire master term, and track whether we've enqueued this term's state update or not as a field within there.

**Comment by DaveCTurner** [comment]
I pushed a sketch of an idea to https://github.com/elastic/elasticsearch/compare/main...DaveCTurner:elasticsearch:2023/10/03/GatewayService-ideas (not tested or even necessarily fully-formed, but maybe it is useful)

File: server/src/main/java/org/elasticsearch/gateway/GatewayService.java [hunk 6]
```diff
@@ -141,83 +144,162 @@ public void clusterChanged(final ClusterChangedEvent event) {
...
+            } else {
+                if (scheduledRecovery == null) {
+                    logger.info(
+                        "delaying initial state recovery for [{}] of term [{}]. expecting [{}] data nodes, but only have [{}]",
+                        recoverAfterTime,
+                        expectedTerm,
+                        expectedDataNodes,
+                        currentDataNodeSize
+                    );
+                    scheduledRecovery = threadPool.schedule(getScheduleTask(), recoverAfterTime, threadPool.generic());
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/gateway/GatewayService.java:230]: @DaveCTurner There are still some complexities with this block of code and other related areas.
1. This class has no state to remember that it has attempt to recover immediately. Therefore if the cluster can recover immediately, it can potentially submit multiple cluster update tasks. Is it a problem?
2. For the scheduled case, we can mostly avoid multiple schedules by checking whether `scheduledRecovery` is `null`. There can still be edge cases where we can schedule more than once due to racing between checking `scheduledRecovery` and reset it back to `null`. If submitting multiple update tasks isn't an issue, we may also chose to not check it all and just always schedule?
3. Because we need to reset `scheduledRecovery` back to null in the scheduled runnable, it needs to be made `volatile` as well.

Do we need to address the 1st point? If so, it seems we need another state variable for it. I forgot to mention it during the sync but this was one of the original complexity. Also because the `ClusterStateUpdateTask` may not run inside `execute` due to dataNodeSize dropping again, the state needs to be reset from within the task which brings back the need of passing a "runAfter" into the task. To simplify things, I think we don't want to check dataNodeSize again inside the task. It's an edge case anyway and dropping it makes things simpler. But we will still need some other state management if we want to address the 1st point. What do you think?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/gateway/GatewayService.java:230]: 1. Good point. I think we should only ever submit one cluster state update task per term, so we ought to track this with a flag within the per-term state.

2. I would not expect any races here, or rather I think if we keep track of whether we've submitted the cluster state update task then that solves those races.

3. Good point, although since we only do that when actually submitting the task again I think the solution is to make this submission a once-only thing.

Comment (ywangd) [reply server/src/main/java/org/elasticsearch/gateway/GatewayService.java]: Pushed 25bbebaa76e715918f613c3666827a9248f0f530 to add a new state variable (`AtomicBoolean`) solved multiple issues that I had. Thanks!

Please let me know if the main code looks good to you.  I'll proceed to add some more tests if you are happy with the main code changes. Please also let me know if you have any ideas for what kinda tests we might need. Thanks!

Comment (ywangd) [reply server/src/main/java/org/elasticsearch/gateway/GatewayService.java]: I have now added multiple tests to cover different scenarios in bf8f21ac0d2145299a94d003eb9909732eb885f2
Now the whole thing looks ready to me.

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Review by DaveCTurner — COMMENTED** [review]
The implementation looks good now, but I do not like the mocks-heavy tests you've added. Tests like this are very fragile and tend to obstruct other refactorings in the same area. There's no need for mocks here, we can create realistic versions of all the dependencies in ways that let us verify that this component really does run its tasks at the right time and not otherwise through their stable APIs.

I pushed [a rough commit which shows how to do that](https://github.com/ywangd/elasticsearch/compare/es-89310-gateway-service...DaveCTurner:elasticsearch:2023/10/07/review-ywangd/es-89310-gateway-service). It doesn't cover everything and probably deserves some more abstraction to split it into multiple test cases but hopefully it's a useful start.

This pattern of creating a `ClusterService` that uses a deterministic threadpool with fake clock is used in a few other places too:

- `org.elasticsearch.cluster.InternalClusterInfoServiceSchedulingTests#testScheduling`
- `org.elasticsearch.cluster.routing.allocation.allocator.DesiredBalanceShardsAllocatorTests#testAllocate`
- `org.elasticsearch.snapshots.SnapshotResiliencyTests.TestClusterNodes.TestClusterNode#TestClusterNode`

They're not all quite the same but it feels like this should be something we can move into `ClusterServiceUtils`.

File: server/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java [hunk 7]
```diff
@@ -38,8 +38,8 @@ public interface ClusterStateTaskExecutor<T extends ClusterStateTaskListener> {
      * already have become master and updated the state in a way that would be inconsistent with the response that {@code N} sends back to
      * clients.
      *
-     * @return The resulting cluster state after executing all the tasks. If {code batchExecutionContext.initialState()} is returned then no
-     * update is published.
+     * @return The resulting cluster state after executing all the tasks. If {@code batchExecutionContext.initialState()} is returned then
+     * no update is published.
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java:42]: The real change here is just adding the missing `@` sign. The other change is due to cascading line wrap.

File: server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java [hunk 8]
```diff
@@ -88,4 +141,355 @@ public void testRecoverStateUpdateTask() throws Exception {
...
+        // recovery is scheduled but has not run yet
+        assertThat(deterministicTaskQueue.hasDeferredTasks(), is(true));
+        assertClusterStateBlocks(true);
+        assertThat(rerouteCount.get(), equalTo(0));
+        final GatewayService.PendingStateRecovery pendingStateRecoveryOfInitialTerm = gatewayService.currentPendingStateRecovery;
+        assertThat(pendingStateRecoveryOfInitialTerm, notNullValue());
+
+        // The 1st scheduled recovery may or may not run depend on what happens to the cluster next
+        final int caseNo = randomIntBetween(0, 2);
+        switch (caseNo) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java:278]: Could we test these cases as separate tests (so they all run every time)?

File: server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java [hunk 9]
```diff
@@ -88,4 +141,355 @@ public void testRecoverStateUpdateTask() throws Exception {
...
+        deterministicTaskQueue.advanceTime();
+        deterministicTaskQueue.runAllRunnableTasks();
+
+        // The 2nd scheduled recovery when data nodes are above recoverAfterDataNodes again
+        deterministicTaskQueue.scheduleAt(
+            initialTimeInMillis + elapsed.millis() * 2,
+            () -> setDataNodeCountTaskQueue.submitTask(randomAlphaOfLength(5), new SetDataNodeCountTask(recoverAfterNodes), null)
+        );
+        deterministicTaskQueue.advanceTime();
+        deterministicTaskQueue.runAllRunnableTasks();
```
Comment (ywangd) [inline server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java:385]: These `advanceTime` then `runAllRunnableTasks()` can be simplified if we have some method like `runToTime(long)` that runs all tasks up to the specified time. I can have a follow-up PR to add this method if you think it is useful.

**Review by DaveCTurner — COMMENTED** [review]
Thanks Yang much nicer IMO. I left one small request about the tests, it's not essential but would be nice if you can find a way to do it without masses of duplication.

File: server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java (hunk 9 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java:385]: Yep I can see value in that, please do.

File: server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java (hunk 8 above)
Comment (ywangd) [reply server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java]: Yeah that's fair. I pushed 19f89bac90000265796a6b551e56049cb5cea881

**Comment by ywangd** [comment]
It is no kidding that this issue is a great source for learning. I spent quite sometime to read through the tests you proposed and related production code. It felt productive. Thanks!

The PR is now updated accordingly. I re-organized the tests for different test purposes and chose to advace the time manually in some cases for testing cancellation. But otherwise the tests are mostly what you have suggested.

**Review by DaveCTurner — COMMENTED** [review]
One more small observation about the tests (I could be persuaded that it's not a blocker)

File: server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java [hunk 10]
```diff
@@ -8,77 +8,131 @@
...
+        rerouteCount = new AtomicInteger();
+        dataNodeIdPrefix = randomAlphaOfLength(10) + "-";
+    }
+
+    private GatewayService createService(final Settings.Builder settingsBuilder, final ClusterState initialState) {
+        final var threadPool = deterministicTaskQueue.getThreadPool();
+        final var settings = settingsBuilder.build();
+        final var clusterSettings = createBuiltInClusterSettings(settings);
+
+        clusterService = new ClusterService(
```
[the remark this answers is not in the document]
Comment (ywangd) [reply server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java]: Sure I pushed b5da2d1edcb1c5dd617cbf5ee191e09007cb7690

An easier alternative is to access `ClusterService` from `GatewayService`. But it requires making the `GatewayService.clusterService` field package private (or add a package private accessor method). Though this pattern is used in many places for testing, I am not sure whether you'd like it. So I did not take this approach. Please let me know whether this would work for you since it might be useful in future.

**Review by DaveCTurner — APPROVED** [review]
LGTM
